design/036-package_id.md

```diff
@@ -0,0 +1,107 @@
+   - Bumping the "minor" version, means that some change was done in the public interface of the library. Maybe some headers that contain templates, or any other thing that could be inlined from headers in the consumers. This "minor" version bump means that the consumers need to be re-built, but in theory they won't break, and it will be possible to recompile them without changing the consumers source code.
+   - Bumping the "major" version means that breaking changes were done in the public interface of the library that will break consumers. This change requires not only to rebuild consumers, but also possibly to change the source code of those consumers to match the new interface.
+- ``core.package_id:default_embed_mode``, default=``full_mode``. This will produce different ``package_id`` when anything in the dependency version, down to the ``package_id`` of the dependency changes. The dependency will be mapped in the consumer ``package_id`` as ``pkg/version@user/channel#recipe_revision:package_id``. That means that if something in the recipe or dependency source code changes, the ``recipe_revision`` will change, and will produce a new consumer ``package_id``, requiring a rebuild. But also if the dependency ``package_id`` changes, maybe because an ``option`` that affects its binary is changed, then the consumer also needs to get a new ``package_id`` and a new binary that is compiled against that different dependency binary.
+
+
+The above is for the regular ``requires``. But other types of requires can also affect the ``package_id``, with the following defaults:
+
+- ``core.package_id:default_python_mode``, default=``minor_mode`` defines how ``python_requires`` will affect the ``package_id`` of the consumer using that ``python_require``. Note that the default is the - ``minor_mode``, because it can use the above versioning flow to represent "no need to build"=patch bump, "needs to rebuild, but won’t break code"=minor bump, "needs to rebuild, will likely break code"=major bump.
+- ``core.package_id:default_build_mode``, default=``None``. By default the ``requires`` in the "build" context (``tool_requires``, ``build_requires``), do not affect their consumers ``package_id``. But with this configuration it can be done.
+
+
+To summarize, things that can be configured (defined in ``global.conf``):
+- ``core.package_id:default_embed_mode``
+- ``core.package_id:default_non_embed_mode``
+- ``core.package_id:default_unknown_mode``
+- ``core.package_id:default_python_mode``
+- ``core.package_id:default_build_mode``
+
+Values that these configurations can take (given the full version is ``1.2.3``):
+- ``full_mode``: including ``dep/1.2.3@user/channel#recipe_revision:package_id``
+- ``major_mode``: include ``dep/1.Y.Z@user/channel``
+- ``minor_mode``: include ``dep/1.2.Z@user/channel``
```

Comment:
I think `minor_mode` should not be affected by the `user/channel` combination. This is now the case, but I think it's confusing. Notably, if you switch `opencv/4.6.0` with `opencv/4.6.0@company/stable`, your packages should not get rebuilt - it's still the same version of OpenCV, just binary from a different provider (possibly built with different flags, but the API and ABI should remain the same).

I still think we need to have a mode that "locks" to the `user/channel` combination or maybe `user` only, but in general that should be a more specific case for specific company rules.

I'm aware that something like this can be done at the [individual package level](https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#versioning-schema) or even shared via python requirements, but, at least from the documentation, I'm not sure that this is as flexible as builtin package modes (i.e. how would you define a custom "semantic version-like" mode that behaves just like the built-in, but ignores the `user/package` combination?).

Maybe I'm just misunderstanding something?

Reply:
We explicitly added this recently after discussing this. We have been considering a few things:
- Replacement (override), of an upstream dependency based on user/channel is something that shouldn't be common, but an exceptional thing. If an ``opencv/4.6.0@company/stable`` dependency is to be used, the recommended approach is to make it explicit in recipes with ``requires = "opencv/4.6.0@company/stable``.
- The idea, once that packages are considered "immutable" and the ``conan copy`` command has been removed in 2.0, is that something under a different user/channel can be a completely different thing, not something to be transparently swapped. Conan 2.0 even enable the ``pkg/version@user``, to drop only the channel.
- Once that we consider packages under different user/channel like a completely different thing, we thought that the least that can happen if that changes is to force by default a re-build of the consumers.

Adding new modes that discard user/channel part is definitely possible, lets see what others think about this default.

Reply:
> Adding new modes that discard user/channel part is definitely possible, lets see what others think about this default.

in our company, we use snapshot/unstable/stable as channel depending on the maturity of the package. Having a mode that only considers `pkg/version@user` would be quite useful for us to avoid unnecessary rebuilds.